README.md
regex-markup
============

regular expression-based text markup

regex-markup is copyright (C) 2001-2005 Oskar Liljeblad.

**version:** 0.10.0 - 2005-05-22

## Introduction


regex-markup performs regular expression-based text markup according to
used-defined rules. This can be used to color syslog files as well as the
output of programs such as ping, traceroute, gcc etc. The hierarchal
rules-format is simple to understand, yet very powerful. Regex-markup is
written entirely in C, and uses POSIX regular expressions to do much of its
work.

For example, these rules can be used to color all lines containing 'kernel:
' green, and in those lines the word 'panic' red:

<pre>
style green { pre "\E[32m" post "\E[39m" }
style red { pre "\E[31m" post "\E[39m" }

/kernel: (.*)/ {
  green
  /panic/ig { red }
}
</pre>

Example rules for `ping`, `traceroute`, and `syslogs` are included.

regex-markup is written by Oskar Liljeblad, and is free software licensed
under terms of the GNU General Public License.

## Copyright and License

<pre>
regex-markup is copyright (C) 2001-2005 Oskar Liljeblad.

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc., 59
Temple Place, Suite 330, Boston, MA  02111-1307  USA

The source code of this project may contain files from other projects, and
files generated by other projects, including:

 * [GNU Autoconf](http://www.gnu.org/software/autoconf/)
 * [GNU Automake](http://sources.redhat.com/automake/)
 * [GNU gettext](http://www.gnu.org/software/gettext/)
 * [Gnulib](http://www.gnu.org/software/gnulib/)
 * [Flex](http://www.gnu.org/software/flex/)
 * [GNU Bison](http://www.gnu.org/software/bison/)
 * [Autoconf Macro Archive](http://ac-archive.sourceforge.net/)

Such files are licensed under the terms of the GNU General Public License
or a license compatible with the GNU GPL (as listed on
http://www.gnu.org/licenses/license-list.html). See each file for copyright
details.

The Swedish translations in the po directory may contain translations from
other projects, including:

 * [GNU C Library](http://www.gnu.org/software/libc/libc.html)
 * [GNU Core Utilities](http://www.gnu.org/software/coreutils/)

See the specific message file (PO file) for copyright of those messages.
</pre>

## Requirements

The following programs are useful (but not required) when building
regex-markup:

 * Bison (or Yacc)
   
    For generating `src/script.c` if it hasn't been generated already.

The following programs are required to build regex-markup:

 * gcc 2.95 or later.
   
    regex-markup uses some features specific to gcc 2.95 (such as macros with variable number of arguments).

The following programs are required for proper operation of regex-markup:

 *  GNU libc6 2.0 or later
   
    regex-markup uses GNU-specific extensions to the C library.

## Installation

Compile regex-markup by running configure and then make. This should produce executables in the src directory, which can be used right away.

If you want to install the regex-markup on your system, run make install.
This will copy the executables and the man pages into the appropriate
directories (usually `/usr/local/bin` and `/usr/local/share/man/man1`,
respecively).

For further details regarding configure and make, see the INSTALL document.

## Common problems

Q: Remark prints this error when started

   remark: Cannot allocate memory

A: Try recompiling regex-markup without the -O2 option, like this:

   `CFLAGS=-g ./configure ; make clean ; make`

Please write me if that doesn't help.

## Usage

regex-markup consists of two programs, remark and cacheline. The main
program, remark, reads lines of text from standard in, highlights them
according to certain rules defined by the user, and writes the highlighted
line to standard out. It only takes one argument - a file containing the
highlighing rules. The format of these files are described in the manual
page of remark.

Examples:

```bash
   ping www.yahoo.com | remark ../examples/rules/ping
   traceroute www.kernel.org | remark ../examples/rules/traceroute
```

The remark program supports the usual long option `--help` which displays
usage and option information. A manual page which can be viewed with

```bash
   man ./remark.1
```

is provided. (A manual page for `cacheline`, cacheline.1, is also provided.)

The cacheline program is used by the logwatcher example program to delay
newlines in output. See the cacheline manpage for more details.

## Format of rule files

See the provided manual page, `remark.1`.

## Homepage

The latest version of regex-markup can be downloaded from

   [http://www.nongnu.org/regex-markup/](http://www.nongnu.org/regex-markup/)

## Feedback

The author of regex-markup and this document, Oskar Liljeblad, can be
reached via email:

   [oskar@osk.mine.nu](mailto:oskar@osk.mine.nu)

Please send bug reports, suggestions, ideas or comments in general to me.

## Known bugs

None at the moment.

## Possible things to do

Please see the TODO document.

-


NEWS
2005-05-22: regex-markup 0.10.0 released.
  Localization support added.
  Use gnulib for improved compatibility.
  Only use lstat64 when available (reported by Michal Fojtik).
  Tested on FreeBSD 4.10 (thanks to Michal Fojtik).
  Update manual pages, add copyright and bug reporting address.
  Cleaned up RPM package spec file.
  Added full GNU GPL header to source files and manual pages.
  Support --help and --version in cacheline.
  Lexer code now includes config.h before everything else.

2003-04-11: regex-markup 0.9.0 released.
  Implemented new command 'set', to simplify substitution constructs.
  Upgraded to autoconf 2.57 and automake 1.7.3.
  Fixed bug that would cause remark to die with "cannot allocate memory"
  when using $n where n is out of range in substitution rules.
  Added new testsuites bug05, e82, e83, e84, e85, e86.
  Clarified README about rule format location.
  Never ignore ctrl+c/SIGINT - I don't remember the benefit of ignoring it
  in the first place. If someone knows, please mail me. :)

2002-01-06: regex-markup 0.8.0 released.
  gcc 2.95 is now required to compile regex-markup.
  Fixed a bug with /(..)*/ matches that would cause regex-markup to
  highlight incorrectly or die ("Cannot allocate memory").
  Fixed a bug in s/../../ rules that would produce wrong output.
  Added example rules for make.
  Improved syslog highlighting rules.

2001-12-21: regex-markup 0.7.0 released.
  Added section about common problems to the README file.
  Remark now also looks in $HOME/.remark and /usr/share/regex-markup for
  rule files.
  The example rule files are installed in /usr/share/regex-markup by
  default.
  Added files for making an RPM package of regex-markup.

2001-12-19: regex-markup 0.6.0 released.
  Cleaned up code to reduce unnecessary memory allocations.
  Added new example rules for diff.
  Rewrote the cacheline program has in C.
  Fixed how escaped characters are dealt with in substitution strings.
  Improved terminal width detection in logwatcher.

2001-12-17: regex-markup 0.5.0 released.
  Almost all code has been rewritten and improved, resulting in more
  features, faster execution, cleaner and less buggy code.
  Implemented line wrapping in remark (previously this was done
  separately). As a result of this, foldlog was removed and logwatcher
  updated to use the build-in wrapping.
  Added a bunch of test cases (to a total of 558 tests).

2001-12-03: regex-markup 0.4.0 released.
  Began support for Cygwin by adding compatibility files from fileutils.
  Updated copyright.
  Added a new program "foldlog" to fold syslog lines.
  Added a testsuite for foldlog.
  Added missing file bison.simple.fixed (needed for complete rebuild).
  Improved error messages.
  Improved the syslog highlighter (logwatcher).
  Improved memory management (don't allocate and free as much memory).
  Fixed a bug in the parsing code that would cause freed memory to be read.

2001-10-28: regex-markup 0.3.0 released.
  Updated manual page.
  Fixed a substitution bug that would cause segmentation violation.
  Fixed style bugs causing incorrect highlighting.
  Better signal handling.
  Improved substitution with support for $`, $&, and $' variables.

2001-10-25: regex-markup 0.2.0 released.
  Added manual pages for remark.
  Added files for making a regex-markup Debian package.
  The author has changed email-address to oskar@osk.mine.nu.

2001-08-13: regex-markup 0.1.0 released.
  First public release.


src/remark.c
/* remark.c - Option-parsing and main function for regex-markup.
 *
 * Copyright (C) 2001-2005 Oskar Liljeblad
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#if HAVE_CONFIG_H
#include <config.h>
#endif
/* POSIX */
#include <unistd.h>
/* C89 */
#include <signal.h>
#include <stdlib.h>
#include <string.h>
/* gnulib */
#include <getopt.h>
#include "dirname.h"
#include "progname.h"
#include "getline.h"
#include "version-etc.h"
/* Gettext */
#include <gettext.h>
#define _(String) gettext(String)
/* common */
#include "common/strbuf.h"
#include "common/string-utils.h"
#include "common/io-utils.h"
#include "common/error.h"
#include "common/llist.h"
#include "common/intutil.h"
/* regex-markup */
#include "remark.h"

#define PROGRAM "remark"

enum {
	VERSION_OPT	= 1000,
	HELP_OPT,
};

static uint32_t prepend_len = 0;
static uint32_t append_len = 0;
static char *prepend_string = "";
static char *append_string = "";

static char *short_opts = "a:f:p:r:w:";
static struct option long_opts[] = {
	{ "prepend",   required_argument, NULL, 'p' },
	{ "append",    required_argument, NULL, 'a' },
	{ "retain",    required_argument, NULL, 'r' },
	{ "width",     required_argument, NULL, 'w' },
	{ "wrap",      required_argument, NULL, 'f' },
	{ "version",   no_argument,	  NULL, VERSION_OPT },
	{ "help",      no_argument,	  NULL, HELP_OPT },
	{ 0 },
};

static void display_help(void);
static void display_version(void);

const char version_etc_copyright[] = "Copyright (C) 2004-2005 Oskar Liljeblad";

static void
display_version(void)
{
	version_etc(stdout, PROGRAM, PACKAGE, VERSION, "Oskar Liljeblad", NULL);
}

static void
display_help(void)
{
	printf(_("Usage: %s [OPTION]... FILE [TEXT]\n\
Read lines of text from standard in (or use TEXT as input string), highlight\n\
according to rules defined in FILE, and print to standard out.\n\
\n\
  -p, --prepend=STRING       string to prepend to all split lines (except last)\n\
  -a, --append=STRING        string to append to all split lines (except first)\n\
  -r, --retain=COUNT         copy characters from first line to wrapped ones\n\
  -w, --width=COLUMNS        wrapping width\n\
  -f, --wrap=TYPE            specifies wrapping type (word/char/none).\n\
      --help                 display this help and exit\n\
      --version              output version information and exit\n\
\n\
Report bugs to <%s>.\n"), program_name, PACKAGE_BUGREPORT);
}

static void
try_line(RemarkScript *script, RemarkInput *input, char *text)
{
	strbuf_set(input->mb.buffer, text);
	input->mb.bufferlen = strlen(text);
	strbuf_set(input->append_mb.buffer, append_string);
	input->append_mb.bufferlen = append_len;
	strbuf_set(input->prepend_mb.buffer, prepend_string);
	input->prepend_mb.bufferlen = prepend_len;

	if (execute_script(script, input)) {
		fputs(strbuf_buffer(input->mb.buffer), stdout);
		fflush(stdout);
	}
}

int
main(int argc, char **argv)
{
	struct sigaction action;
	RemarkInput input;
	int c;

	set_program_name(argv[0]);

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	memset(&action, 0, sizeof(struct sigaction));
	/* 2003-04-11 : I don't know why this is good in the first place... */
	/*if (!isatty(STDIN_FILENO)) {
		action.sa_handler = SIG_IGN;
		action.sa_flags = SA_RESTART;
		if (sigaction(SIGINT, &action, NULL) < 0)
			die_errno(NULL);
	}*/

	while ((c = getopt_long (argc, argv, short_opts, long_opts, NULL)) != -1) {
		switch (c) {
		case 'p':
			prepend_string = optarg;
			break;
		case 'a':
			append_string = optarg;
			break;
		case 'r':
			if (!parse_uint32(optarg, &wrap_retain))
				die(_("invalid retain value: %s"), optarg);
			break;
		case 'w':
			if (!parse_uint32(optarg, &wrap_width))
				die(_("invalid width: %s"), optarg);
			break;
		case 'f':
			if (!identify_wrapper(optarg))
				die(_("invalid wrapping type: %s"), optarg);
			break;
		case VERSION_OPT:
			display_version();
			exit(0);
			break;
		case HELP_OPT:
			display_help();
			exit(0);
			break;
		case '?':
			exit(1);
		}
	}

	append_len = strlen(append_string);
	prepend_len = strlen(prepend_string);
	if (wrapper == 0)
		wrapper = WRAPPER_CHAR;
	if (wrap_retain + append_len + prepend_len >= wrap_width)
		die(_("retain, append and prepend too long"));

	init_input(&input);

	if (argc-optind >= 2) {
		RemarkScript *script = parse_script(argv[optind]);
		for (c = optind+1; c < argc; c++)
			try_line(script, &input, argv[c]);
		free_script(script);
	}
	else {
		RemarkScript *script = NULL;
		size_t linesize = 0;
		char *line = NULL;

		script = parse_script(argc > optind ? argv[optind] : NULL);
		while (getline(&line, &linesize, stdin) != -1) {
			chomp(line);
			try_line(script, &input, line);
		}
		free(line);
		free_script(script);

		if (ferror(stdin))
			die_errno(_("cannot read from standard in"));
	}

	free_input(&input);

	exit(0);
}
